File: Fsharp/TreeGenerator.fs
module Fsharp.TreeGenerator

type tree =
    | Node of left : tree * right : tree * numberOfLeaves : int
    | Leaf of int64

type TreeGenerator() =
    let rnd = System.Random()

    member this.CreateBinaryTree depth valueGenerator =
        match depth with
        | 1 -> Leaf (valueGenerator())
        | n -> Node(
                   left = (this.CreateBinaryTree (depth - 1) valueGenerator),
                   right = (this.CreateBinaryTree (depth - 1) valueGenerator),
                   numberOfLeaves = int (2.0f ** (float32 depth)))

    member this.CreateTree maxChildren valueGenerator =
        if maxChildren = 1 then
            Leaf(valueGenerator())
        else 
            let nNodesLeft = rnd.Next(1, maxChildren)
            let nNodesRight = maxChildren - nNodesLeft
            let l = this.CreateTree nNodesLeft valueGenerator
            let r = this.CreateTree nNodesRight valueGenerator
            Node (l, r, maxChildren)

File: LenientCsharp/TreeGenerator.cs
using System;

namespace LenientBenchmark
{
    internal class TreeGenerator
    {
        private static readonly Random rnd = new Random();

        public static Tree<T> CreateTree<T>(int maxChildren, Func<T> valueGenerator)
        {
            if(maxChildren == 1)
                return new Leaf<T>(valueGenerator());
            
            // Distribute remaining children randomly over the two subtrees
            var childrenLeft = rnd.Next(1, maxChildren);
            var childrenRight = maxChildren - childrenLeft;
            
            var l = CreateTree(childrenLeft, valueGenerator);
            var r = CreateTree(childrenRight, valueGenerator);
            
            return new Node<T>(l, r, maxChildren);
        }

        public static Tree<T> CreateBinaryTree<T>(int depth, Func<T> valueGenerator)
        {
            if (depth == 1)
                return new Leaf<T>(valueGenerator());
            return new Node<T>(
                CreateBinaryTree(depth - 1, valueGenerator), CreateBinaryTree(depth - 1, valueGenerator), (int) Math.Pow(2, depth));
        }
    }
}